core/src/main/java/org/apache/calcite/sql/SqlRangeN.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.sql;
+
+import org.apache.calcite.sql.parser.SqlParserPos;
+
+import java.util.List;
+
+public class SqlRangeN extends SqlCall{
+
+  final SqlNodeList rangeList;
+  final SqlNode testIdentifier;
+  final NoRangeUnknown extraPartitions;
+
+  public SqlRangeN(final SqlParserPos pos, final SqlNode testIdentifier
+      , final SqlNodeList rangeList, final NoRangeUnknown extraPartitions) {
+    super(pos);
+    this.testIdentifier = testIdentifier;
+    this.rangeList = rangeList;
+    this.extraPartitions = extraPartitions;
+  }
+
+  @Override public SqlOperator getOperator() {
+    return null;
+  }
+
+  @Override public List<SqlNode> getOperandList() {
+    return null;
+  }
+
+  @Override public void unparse(final SqlWriter writer, final int leftPrec
+      , final int rightPrec) {
+    writer.keyword("RANGE_N");
+    SqlWriter.Frame funcCallFrame = writer.startList(SqlWriter.FrameTypeEnum.FUN_CALL
+        , "(", ")");
+    testIdentifier.unparse(writer, leftPrec, rightPrec);
+    writer.keyword("BETWEEN");
+    SqlWriter.Frame frame = writer.startList(SqlWriter.FrameTypeEnum.SIMPLE
+        , "", "");
+    for (SqlNode range: rangeList.getList()) {
+      writer.sep(",");
+      range.unparse(writer, leftPrec, rightPrec);
+    }
+    if (extraPartitions != null) {
```

Comment:
This logic looks quite similar to the unparsing logic in CASE_N. Should we put in some common location (e.g. a base class' unparse method) and just have the keyword "CASE" or "RANGE" be passed in as a parameter?

Reply:
Yes, the syntax and logic are very similar. They are not only code level similar but also have similar definition. Could you clarify, if you are suggesting we could make a base class for the now enum represented keyword, or are you suggesting we make a base class for SqlRangeN and SqlCaseN?

Reply:
Perhaps we can have a base class for SqlRangeN and SqlCaseN to derive from. And we can make the enums more generic such that it would work for both cases (e.g. NONE, UNKNOWN, NONE_OR_UNKNOWN, NONE_COMMA_UNKNOWN)?

Reply:
I've combined the two enums into one in https://github.com/googleinterns/calcite/pull/212/commits/91c78239f8c8b101d26812dd2cab0fb8fe539db7. After looking deeper into the grammar of Case_N and Range_N, they are similar, but I don't see much benefit as of now there are only two classes Case_N and Range_N. We could always make the decision to abstract the logic out later when there are more alike classes.

Reply:
Ok thanks for combining the enums. Keeping them as separate classes for now SGTM.